src/SignalR/server/Core/src/Internal/DefaultHubDispatcher.cs

```diff
@@ -474,13 +474,15 @@ private Task ProcessStreamItem(HubConnectionContext connection, StreamItemMessag
         catch (ChannelClosedException ex)
         {
             // If the channel closes from an exception in the streaming method, grab the innerException for the error from the streaming method
+            Log.FailedStreaming(_logger, invocationId, descriptor.MethodExecutor.MethodInfo.Name, ex.InnerException ?? ex);
             error = ErrorMessageHelper.BuildErrorMessage("An error occurred on the server while streaming results.", ex.InnerException ?? ex, _enableDetailedErrors);
         }
         catch (Exception ex)
         {
             // If the streaming method was canceled we don't want to send a HubException message - this is not an error case
             if (!(ex is OperationCanceledException && streamCts.IsCancellationRequested))
             {
+                Log.FailedStreaming(_logger, invocationId, descriptor.MethodExecutor.MethodInfo.Name, ex);
```

Comment:
Not sure should this one. cancellation failures shouldn't be logged as an error, maybe debug.

Reply:
It isn't see `!`. Or do you mean other `OperationCanceledException`s when `!streamCts.IsCancellationRequested`?

Reply:
Ah good catch @campersau! I missed the ! too.

For the `ChannelClosedException`, we should have a test too right?

Reply:
There isn't currently any test for the `ChannelClosedException` case, so I have added one which manually throws a `ChannelClosedException`.
Because from looking at the Channels implementation in runtime it looks like these don't throw `ChannelClosedException` from within `WaitToReadAsync` / `TryRead` but instead always throw the users exception from `TryComplete(Exception ex)`.
Only `ReadAsync` would throw a `ChannelClosedException` as far as I can tell.

https://github.com/dotnet/runtime/blob/57bfe474518ab5b7cfe6bf7424a79ce3af9d6657/src/libraries/System.Threading.Channels/src/System/Threading/Channels/BoundedChannel.cs#L208-L210

https://github.com/dotnet/runtime/blob/57bfe474518ab5b7cfe6bf7424a79ce3af9d6657/src/libraries/System.Threading.Channels/src/System/Threading/Channels/UnboundedChannel.cs#L167-L169

For reference `AsyncEnumerableAdapters` only uses `WaitToReadAsync` / `TryRead`:
https://github.com/dotnet/aspnetcore/blob/3ea008c80d5cc63de7f90ddfd6823b7b006251ff/src/SignalR/common/Shared/AsyncEnumerableAdapters.cs#L55

---

Regarding: `Log.FailedInvokingHubMethod`
I can change it to use this instead, but maybe a different log might be better to distinguish between those two cases. The case where it would throw immediately vs. the case where it could throw mid reading. (The tests currently throw directly for the first read but it could also happen much later.)
Also  `Log.FailedInvokingHubMethod` does not include the `invocationId` which might be useful as this exception might be logged much later.
In general the logs are pretty inconsistent with the inclusion of the `invocationId`.

Reply:
Looks like `ChannelClosedException` was relevant many years ago in the first implementation of the feature: https://github.com/dotnet/aspnetcore/blame/1289636c3afa0e5f297df141359e4ba72dac6e70/src/SignalR/common/Shared/AsyncEnumerableAdapters.cs#L28

I'm not against a new log message, it's always helpful to have more specific logs when investigating issues (although you can argue here that the exception stack trace would give you enough information).

I would also be happy with adding the invocation ID to the `FailedInvokingHubMethod` log.

Reply:
Should we remove the `ChannelClosedException` case? I guess there could still be custom Channel implementations which do throw a `ChannelClosedException`, so maybe no?

---

There are currently three proposed options for logging:

1. Use a new logger message like `FailedStreaming` as currently proposed in the PR.
2. Use the existing logger message `FailedInvokingHubMethod`.
3. Use the existing logger message `FailedInvokingHubMethod` and add the `invocationId`.

I am fine with either of those, just wanted to bring up the discussion.

If we want to go with (3) then there are some other logger messages which currently don't have an `invocationId` so we could add it to them too for consistentcy. Here are all of these:

https://github.com/dotnet/aspnetcore/blob/c9bd468c2075678d3b764ff9159c25db17cd7049/src/SignalR/server/Core/src/Internal/DefaultHubDispatcher.cs#L287

https://github.com/dotnet/aspnetcore/blob/c9bd468c2075678d3b764ff9159c25db17cd7049/src/SignalR/server/Core/src/Internal/DefaultHubDispatcher.cs#L305

https://github.com/dotnet/aspnetcore/blob/c9bd468c2075678d3b764ff9159c25db17cd7049/src/SignalR/server/Core/src/Internal/DefaultHubDispatcher.cs#L349

https://github.com/dotnet/aspnetcore/blob/c9bd468c2075678d3b764ff9159c25db17cd7049/src/SignalR/server/Core/src/Internal/DefaultHubDispatcher.cs#L389

https://github.com/dotnet/aspnetcore/blob/c9bd468c2075678d3b764ff9159c25db17cd7049/src/SignalR/server/Core/src/Internal/DefaultHubDispatcher.cs#L395

https://github.com/dotnet/aspnetcore/blob/c9bd468c2075678d3b764ff9159c25db17cd7049/src/SignalR/server/Core/src/Internal/DefaultHubDispatcher.cs#L451

https://github.com/dotnet/aspnetcore/blob/c9bd468c2075678d3b764ff9159c25db17cd7049/src/SignalR/server/Core/src/Internal/DefaultHubDispatcher.cs#L458

https://github.com/dotnet/aspnetcore/blob/c9bd468c2075678d3b764ff9159c25db17cd7049/src/SignalR/server/Core/src/Internal/DefaultHubDispatcher.cs#L593

https://github.com/dotnet/aspnetcore/blob/c9bd468c2075678d3b764ff9159c25db17cd7049/src/SignalR/server/Core/src/Internal/DefaultHubDispatcher.cs#L603

Reply:
> Should we remove the `ChannelClosedException` case?

I don't see any harm in keeping it.

> There are currently three proposed options for logging:

I'm going to vote for adding a new log, it's more user friendly to see specific logs.

We can look at adding invocation IDs to existing logs in a separate change.